File: java/src/test/java/org/craftedsw/tripservicekata/trip/TripServiceTest.java
package org.craftedsw.tripservicekata.trip;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import org.craftedsw.tripservicekata.exception.UserNotLoggedInException;
import org.craftedsw.tripservicekata.user.User;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class TripServiceTest {

    User sandro = new User();
    TripService tripService;

    @Test
    void shouldFailWhenUserIsNotLoggedIn() {
        User notLoggedIn = null;
        createTripServiceFor(notLoggedIn);

        assertThrows(UserNotLoggedInException.class, () -> tripService.getTripsByUser(sandro));
    }

    @Test
    void shouldNotListTripsIfUserIsNotAFriend() {
        User notAFriend = new User();
        createTripServiceFor(notAFriend);

        List<Trip> trips = tripService.getTripsByUser(sandro);

        assertTrue(trips.isEmpty());
    }

    @Test
    void shouldListTripsForFriend() {
        User currentUser = new User();
        sandro.addFriend(currentUser);
        List<Trip> tripsOfSandro = Arrays.asList(new Trip());
        createTripServiceWith(currentUser, tripsOfSandro);

        List<Trip> trips = tripService.getTripsByUser(sandro);

        assertEquals(tripsOfSandro, trips);
    }

    @Test
    void shouldListTripsForFriendLaterInList() {
        User currentUser = new User();
        sandro.addFriend(new User());
        sandro.addFriend(new User());
        sandro.addFriend(currentUser);
        List<Trip> tripsOfUser = Arrays.asList(new Trip());
        createTripServiceWith(currentUser, tripsOfUser);

        List<Trip> trips = tripService.getTripsByUser(sandro);

        assertEquals(tripsOfUser, trips);
    }

    private void createTripServiceFor(User stubbedLoggedInUser) {
        createTripServiceWith(stubbedLoggedInUser, Collections.emptyList());
    }

    private void createTripServiceWith(User stubbedLoggedInUser, List<Trip> stubbedTripsOfUser) {
        tripService = new TripService() {
            @Override
            protected User getLoggedInUser() {
                return stubbedLoggedInUser;
            }

            @Override
            protected List<Trip> findTripsByUser(User user) {
                if (user.equals(sandro)) {
                    return stubbedTripsOfUser;
                }
                Assertions.fail("expected user sandro " + sandro + ", got user " + user);
                return null;
            }
        };
    }

}

File: python-objects/TripServiceTest.py
import unittest

from Trip import Trip
from TripService import TripService
from User import User
from UserNotLoggedInException import UserNotLoggedInException


class TripServiceTest(unittest.TestCase):
    def test_exception_when_not_logged_in(self):
        for_user = User()

        trip_service = TripService()
        try:
            trip_service._get_trips_pure(for_user,
                                         lambda: FakeUserSession(None),
                                         lambda user: None)
            self.fail("Should raise an exception")
        except UserNotLoggedInException:
            pass  # ok

    def test_list_no_trips_when_not_friend(self):
        logged_in_user = User()
        for_user = User()

        trip_service = TripService()
        trips = trip_service._get_trips_pure(for_user,  #
                                             lambda: FakeUserSession(logged_in_user),
                                             lambda user: None)

        self.assertEquals([], trips)

    def test_list_trips_when_friend(self):
        logged_in_user = User()
        for_user = User()
        for_user.add_friend(logged_in_user)
        users_trips = [Trip()]

        trip_service = TripService()
        trips = trip_service._get_trips_pure(for_user,  #
                                             lambda: FakeUserSession(logged_in_user),
                                             lambda user: users_trips)

        self.assertEquals(users_trips, trips)


class FakeUserSession(object):
    def __init__(self, logged_in_user):
        self.logged_in_user = logged_in_user

    def get_logged_user(self):
        return self.logged_in_user


if __name__ == '__main__':
    unittest.main()
